sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/ScriptTransformationExec.scala

```diff
@@ -137,13 +137,12 @@ case class ScriptTransformationExec(
             throw writerThread.exception.get
           }
 
-          if (!proc.isAlive) {
-            val exitCode = proc.exitValue()
-            if (exitCode != 0) {
-              logError(stderrBuffer.toString) // log the stderr circular buffer
-              throw new SparkException(s"Subprocess exited with status $exitCode. " +
-                s"Error: ${stderrBuffer.toString}", cause)
-            }
+          proc.waitFor()
```

Comment:
I tested with provided python script. Actually at ScriptTransformationWriterThread, no exception is thrown. 

And, checkFailureAndPropagate is called not due to EOFException, it is called at:

https://github.com/apache/spark/blob/8a17d26784c53fb50b6373b566aab71135c8956f/sql/hive/src/main/scala/org/apache/spark/sql/hive/execution/ScriptTransformationExec.scala#L163-L167

In this case, checkFailureAndPropagate can be called before proc is being terminated due to possible lag, so the !proc.isAlive check fails.

I think the fix is to add proc.waitFor() before other checkFailureAndPropagate calls.

Reply:
@viirya In the proposed PR, proc.waitFor() is called within function checkFailureAndPropagate. Basically, proc is terminated before any checkFailureAndPropagate call returns. Do you mean that we should not waitfor proc before other CheckFailureAndPropagate calls but the case you mentioned?